metals/src/main/scala/scala/meta/internal/metals/FoldingRangeProvider.scala

```diff
@@ -0,0 +1,122 @@
+package scala.meta.internal.metals
+
+import java.util
+import java.util.Collections
+import org.eclipse.lsp4j.FoldingRange
+import org.eclipse.lsp4j.FoldingRangeCapabilities
+import scala.meta.Template
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta._
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.FoldingRangeProvider._
+import scala.meta.io.AbsolutePath
+import scala.meta.tokens.Token.KwYield
+
+final class FoldingRangeProvider(val trees: Trees, foldOnlyLines: Boolean) {
+  def getRangedFor(path: AbsolutePath): util.List[FoldingRange] = {
+    trees
+      .get(path)
+      .map(findFoldingRanges)
+      .getOrElse(util.Collections.emptyList())
+  }
+
+  private def findFoldingRanges(tree: Tree): util.List[FoldingRange] = {
+    val ranges = new FoldingRanges(foldOnlyLines)
+
+    tree traverse {
```

Comment:
The biggest value in doing folding based on the syntax tree is that we don't have to limit ourselves to tree nodes with curly braces or parentheses, for example

```scala
"""|a
   |b
   |c
   |d
   |e
   |f""".stripMargin
```

Reply:
Just brainstorming, we could fold the first/last line based on the length of the range

- 0-2 lines: no folding (it's small enough we don't care)
- 3-4 lines: fold start and end line
- 5+ lines: fold end line and keep start line

Reply:
That sounds interesting - at least for those cases for which we cannot have a precise folding. I will check it out

Reply:
a generic rule like that (well, not a `Stat` because it meant everything - even class declaration - but a `Term`) gives us somewhat undesired results like:

```
       val manyLines = >>region>>>>region>>"""|a
       |b
       |c
       |d
       |e
       |f"""<<region<<.stripMargin<<region<<
```
so I will have to investigate it further - but in the end it may not be so simple

Reply:
What if we limit the traversal to only fold top-level statements in `Term.Block`, `Template.stats` and `Pkg.stats`?

Reply:
Let me create more test cases handled as best as they can and only then revisit this idea. It seems like something worthwhile to check - but I am not yet confident enough it will not interfere with other constructs.